pkg/controller/gkenetworkparamset/gnpcontroller_validations.go

```diff
@@ -0,0 +1,224 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gkenetworkparamset
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+	"google.golang.org/api/compute/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	networkv1 "k8s.io/cloud-provider-gcp/crd/apis/network/v1"
+)
+
+func updateConditions(conditions []v1.Condition, updatedCondition v1.Condition) []v1.Condition {
+
+	// to be safe, lets not mutate the original conditions
+	newConditions := make([]v1.Condition, len(conditions))
+	copy(newConditions, conditions)
+
+	updated := false
+	for i, existingCondition := range newConditions {
+		if existingCondition.Type == updatedCondition.Type {
+			newConditions[i] = updatedCondition
+			updated = true
+			break
+		}
+	}
+	if !updated {
+		newConditions = append(newConditions, updatedCondition)
+	}
+	return newConditions
+}
+
+type gnpValidation struct {
+	IsValid      bool
+	ErrorReason  networkv1.GKENetworkParamSetConditionReason
+	ErrorMessage string
+}
+
+func (val *gnpValidation) toCondition() v1.Condition {
+	condition := v1.Condition{}
+
+	if val.IsValid {
+		condition.Status = v1.ConditionTrue
+	} else {
+		condition.Status = v1.ConditionFalse
+		condition.Reason = string(val.ErrorReason)
+		condition.Message = val.ErrorMessage
+	}
+
+	condition.Type = string(networkv1.GKENetworkParamSetStatusReady)
+	condition.LastTransitionTime = v1.NewTime(time.Now())
+
+	return condition
+}
+
+func (c *Controller) getAndValidateSubnet(ctx context.Context, params *networkv1.GKENetworkParamSet) (*compute.Subnetwork, *gnpValidation) {
+	if params.Spec.VPCSubnet == "" {
+		return nil, &gnpValidation{
+			IsValid:      false,
+			ErrorReason:  networkv1.SubnetNotFound,
+			ErrorMessage: fmt.Sprintf("subnet not specified"),
+		}
+	}
+
+	// Check if Subnet exists
+	subnet, err := c.gceCloud.GetSubnetwork(c.gceCloud.Region(), params.Spec.VPCSubnet)
+	if err != nil || subnet == nil {
+		fetchSubnetErrs.Inc()
+		return nil, &gnpValidation{
+			IsValid:      false,
+			ErrorReason:  networkv1.SubnetNotFound,
+			ErrorMessage: fmt.Sprintf("%s not found in %s", params.Spec.VPCSubnet, params.Spec.VPC),
+		}
+	}
+
+	return subnet, &gnpValidation{IsValid: true}
+}
+
+func (c *Controller) validateGKENetworkParamSet(ctx context.Context, params *networkv1.GKENetworkParamSet, subnet *compute.Subnetwork) (*gnpValidation, error) {
+
+	// Check if secondary range exists
+	if params.Spec.PodIPv4Ranges != nil && params.Spec.DeviceMode == "" {
```

Comment:
I see you indirectly do this check in cross validation with Network, but we can tell from just GNP that it is wrong, you need one of the 2. The cross validates that we are not referencing e.g. deviceMode with type=L3, so that check is still needed as well.

Reply:
done in gnp validations